runtime.md

```diff
@@ -1,5 +1,47 @@
 # Runtime and Lifecycle
 
+## State
+
+The runtime state for a container is persisted on disk so that external tools can consume and act on this information.
+The runtime state is stored in a JSON encoded file.
+It is recommended that this file is stored in a temporary filesystem so that it can be removed on a system reboot.
+On Linux based systems the state information should be stored in `/run/oci/containers`.
+The directory structure for a container is `/run/oci/containers/<containerID>/state.json`.
+By providing a default location that container state is stored external applications can find all containers running on a system.
+
+* **version** (string) Version of the OCI specification used when creating the container.
+* **id** (string) ID is the container's ID.
+* **pid** (int) Pid is the ID of the main process within the container.
+* **root** (string) Root is the path to the container's bundle directory.
+
+The ID is provided in the state because hooks will be executed with the state as the payload.
+This allows the hook to perform clean and teardown logic after the runtime destroys it's own state.
+The root directory to the bundle is provided in the state so that consumers can find the container's configuration and rootfs.
+
+*Example*
+
+```json
+{
+    "id": "oci-container",
+    "pid": 4422,
+    "root": "/containers/redis"
+}
+```
+
+Linux systems add some platform specific information to the state.
+
+* **externalFds** Paths to the container's open file descriptors.
+
+*Example Linux*
+
+```json
+{
+    "externalFds": [
+        "/proc/33/fd/1"
```

Comment:
On Fri, Aug 28, 2015 at 08:24:59AM -0700, Andrew Vagin wrote:
> ExtraFiles doesn't include stdin, stdout, stderr…

Ok, I can live with that.  So the idea is that externalFds[i] tells
you where to connect the restoring runtime's FD i+2.  That doesn't
cleanly handle cases where we have gaps in the file descriptors:

  # runc config.json 123<foo

I'd still prefer something like:

  "externalFds": {
    "123": "/home/wking/foo",
  }

to:

  "externalFds": [-1, -1, -1, …, -1, "/home/wking/foo"],

to setup that high-valued file descriptor.

I'm also not clear on what the value should be (“/home/wking/foo”
above).  For examples like 123<foo, where we're passing in a plain
file's descriptory, it's reasonably clear (modulo seeks and such).
But how do we handle unnamed pipes?  Is “pipe:[18132]” the read or
write end of that pipe?  What if the client is duping descriptors?
What if we pass a Unix socket in, and then (after start-time), pass
*another* file descriptor down that socket with sendmsg?  What if the
client has a listening socket and has accepted a new connection?

I'm having trouble imagining a system that works for all cases.
That's fine, but I'd like externalFds to land with docs that describes
its intended value clearly, and lays out the logic for deciding
whether or not a particular application's file descriptor
manipulations will break the checkpoint/restore functionality.
Linking to CRIU docs would be fine for the parts of this that are
covered in the CRIU docs.